### PR TITLE
cpu: enable gemm-bf16f32 for SDPA BF16 

### PR DESCRIPTION
**Pull request by aditew01**
This PR enables SDPA BF16:  gemm:bf16f32 for aarch64.  This will enable faster inference for models with attention layers  for autocast mode (bf16).

Benchmark results from  [PyTorch CI HUD - branch](https://hud.pytorch.org/benchmark/huggingface/inductor_no_cudagraphs?dashboard=torchinductor&startTime=Fri%2C%2028%20Mar%202025%2021%3A26%3A20%20GMT&stopTime=Fri%2C%2004%20Apr%202025%2020%3A26%3A20%20GMT&granularity=hour&mode=inference&dtype=bfloat16&deviceName=cpu%20(aarch64)&lBranch=adi/gemm_bf16f32&lCommit=d5aeab452e4b1f0580a4636b15a604c77a02c57b&rBranch=main&rCommit=bc72420bcb37390af3fced885e019903e6e425bd)
Overall Geometric mean speedup in HUD dashboard  : for Huggingface: `[0.48x → 0.58x]` and for Blueberries: `[0.88x → 1.13x]`  

Benchmark numbers for `torch.nn.functional.scaled_dot_product_attention`on Neoverse™ V1. 

`batch_size = 1, num_attention_heads = 64, sequence_length = 512, attention_head_size = 128`
 `threads=16` 
Screenshot 2024-12-20 at 16 23 22

Script to benchmark & profile SDPA: 

    import torch
    import torch.nn as nn
    import time
    import numpy as np
    from torch.profiler import profile, record_function, ProfilerActivity
    class SimpleAttentionModel(nn.Module):
        def __init__(self, query, key, value):
            super(SimpleAttentionModel, self).__init__()
            self.query = query
            self.key = key
            self.value = value

        def forward(self, attn_mask=None):
            torch.nn.functional.scaled_dot_product_attention(
                        self.query,
                        self.key,
                        self.value,
                        attn_mask=attn_mask)

    #batch_size = 1, num_attention_heads = 64, sequence_length = 512, hidden_size = 128
    def bench_sdpa(batch_size = 1, num_attention_heads = 64, sequence_length = 512, query_sequence_length = 128 , hidden_size=128, precision=torch.float32):
        with torch.no_grad():
            attention_head_size = int(hidden_size / num_attention_heads)
            query = torch.rand(size=(batch_size, num_attention_heads, query_sequence_length, attention_head_size), dtype=precision)
            key = torch.rand(size=(batch_size, num_attention_heads, sequence_length, attention_head_size), dtype=precision)
            value = torch.rand(size=(batch_size, num_attention_heads, sequence_length, attention_head_size), dtype=precision)
            
            model = SimpleAttentionModel(query, key, value)
            model.eval()        
            for _ in range(10):
                model()
            times = []
            n_iters = 100
            for _ in range(n_iters):
                s = time.time_ns()
                model()
                times.append((time.time_ns() - s) / 1e3)
            min_times = np.min(times)
            mean_times = np.mean(times)
            print(f"Min Times = {min_times} us")
            print(f"Mean Times = {mean_times} us")
            print("Times = ", times)

    print("BF16 mode:")
    with profile(activities=[ProfilerActivity.CPU], record_shapes=True) as prof:
        with record_function("model_inference"):
            bench_sdpa(precision=torch.bfloat16)
    profile_data = prof.key_averages(group_by_input_shape=True).table(sort_by="cpu_time_total")
    print(profile_data)

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @gujinghui @PenghuiCheng @jianyuh @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal @malfet @milpuz01 @nikhil-arm @fadara01 @voznesenskym @penguinwu @EikanWang @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @xmfan

### CODE REVIEW

**Comment by aditew01** [comment]
cc: @malfet @jgong5

File: aten/src/ATen/native/CPUBlas.cpp [hunk 1]
```diff
@@ -402,6 +405,13 @@ void gemm(
     const float beta,
     float *c, int64_t ldc) {
   internal::normalize_last_dims(transa, transb, m, n, k, &lda, &ldb, &ldc);
+
+// TODO: ADI: add heuristic based on shape to dispatch to sbgemm_ vs MKLDNN
+#if AT_MKLDNN_ENABLED()
+   if (mkldnn_bf16f32_gemm(transa, transb, m, n, k, alpha, a, lda, b, ldb, beta, c, ldc)) {
```
Comment (jgong5) [inline aten/src/ATen/native/CPUBlas.cpp:411]: Since this change is not only for aarch64 but also impact x86, the major concern is the MKLDNN gemm would be slower than MKL's. Some options for you to consider:
1. Move MKL call before this one so that we do not regress x86.
2. Only apply MKLDNN gemm for aarch64 in this PR so that we do not regress x86.
3. Do more benchmarking to prove that there is no regression for x86.

Comment (aditew01) [reply aten/src/ATen/native/CPUBlas.cpp:411]: I've updated the patch @jgong5 . 
I'm guarding it with aarch64 for now.

File: aten/src/ATen/native/CPUBlas.cpp [hunk 2]
```diff
@@ -402,6 +405,13 @@ void gemm(
     const float beta,
     float *c, int64_t ldc) {
   internal::normalize_last_dims(transa, transb, m, n, k, &lda, &ldb, &ldc);
+
+#if defined(__aarch64__) && AT_MKLDNN_ENABLED()
```
Comment (jgong5) [inline aten/src/ATen/native/CPUBlas.cpp:409]: Do you mind add a TODO here to support non-aarch64 CPUs in the future?

**Review by jgong5 — APPROVED** [review]

File: aten/src/ATen/native/CPUBlas.cpp [hunk 3]
```diff
@@ -322,6 +322,24 @@ void gemm(
    const float beta,
    at::BFloat16 *c, int64_t ldc) {
    internal::normalize_last_dims(transa, transb, m, n, k, &lda, &ldb, &ldc);
+
```
Comment (swolchok) [inline aten/src/ATen/native/CPUBlas.cpp:325]: I don't understand how this change is related to the stated purpose of this PR. the rest of the change appears to be in an entirely different function.

File: aten/src/ATen/native/mkldnn/Matmul.cpp [hunk 4]
```diff
@@ -190,6 +189,82 @@ mkldnn_gemm(
...
+    float alpha,
+    const scalar_t *a_data, int64_t lda,
+    const scalar_t *b_data, int64_t ldb,
+    float beta,
+    float* c_data, int64_t ldc) {
+// TODO: ADI
+// introduce heuristic to validate dispatch to MKLDNN
+// (m * n * k <= 16 * 16 * 16)
+  bool bf16_usable = std::is_same_v<scalar_t, c10::BFloat16> && use_mkldnn_bf16_matmul();
+  if ( !(bf16_usable) ) {
```
Comment (malfet) [inline aten/src/ATen/native/mkldnn/Matmul.cpp:208]:
```suggestion
  if (!bf16_usable) {
```

File: aten/src/ATen/native/CPUBlas.cpp (hunk 3 above)
Comment (malfet) [reply aten/src/ATen/native/CPUBlas.cpp:325]: Agree with the above, how this is related to enabling gemm-bf16f32?

**Review by malfet — CHANGES_REQUESTED** [review]
LGTM, but please don't change the order of operations in `gemm` function

File: aten/src/ATen/native/CPUBlas.cpp (hunk 3 above)
Comment (Ryo-not-rio) [reply aten/src/ATen/native/CPUBlas.cpp:325]: Replying on behalf of @aditew01 : `mkldnn_gemm` is moved to before the `blas_gemm` as the mkldnn implementation is faster than the blas implementation for aarch64

Comment (swolchok) [reply aten/src/ATen/native/CPUBlas.cpp:325]: that doesn't sound related to this PR.

Comment (aditew01) [reply aten/src/ATen/native/CPUBlas.cpp]: Sounds reasonable. I've reverted the change and will open a new PR to address this.

File: aten/src/ATen/native/mkldnn/Matmul.cpp (hunk 4 above)
Comment (aditew01) [reply aten/src/ATen/native/mkldnn/Matmul.cpp]: ACK. I've made the suggested change.

**Comment by aditew01** [comment]
@malfet @swolchok request for a fresh review. I've addressed the review comments.

File: aten/src/ATen/native/mkldnn/Matmul.cpp [hunk 5]
```diff
@@ -154,22 +153,131 @@ mkldnn_gemm(
   if constexpr (std::is_same_v<scalar_t, float>) {
     idtype = ideep::tensor::data_type::f32;
   }
+  #ifdef __aarch64__
```
Comment (swolchok) [inline aten/src/ATen/native/mkldnn/Matmul.cpp:156]: why is it correct to just ignore strides on aarch64?

**Review by swolchok — COMMENTED** [review]
I don't really know anything about MKLDNN and I don't think I have authority to approve PyTorch diffs anyway, but this was in my notifications so here's a question.

File: aten/src/ATen/native/mkldnn/Matmul.cpp (hunk 5 above)
Comment (aditew01) [reply aten/src/ATen/native/mkldnn/Matmul.cpp:156]: oneDNN is free to reorder the weights into an optimized format when strides are not passed. 
When passing strides to MKLDNN, we can enforce a blocking format. This can affect the kernel selection. 

Passing strides in oneDNN (MKLDNN) enforces a blocking format, which can affect kernel selection. Without strides, oneDNN can **reorder weights into an optimized format**, improving performance.

#### **Benchmark Insights**
Using [`benchdnn`](https://github.com/oneapi-src/oneDNN/blob/main/tests/benchdnn/README.md), I observed different kernel mappings based on whether strides were passed.

#### **With Strides (`768x1`)**
```sh
ONEDNN_VERBOSE=all ./benchdnn --matmul --dt=bf16:bf16:f32 --strides=768x1 14x64:64x142

Maps to gemm:jit:bf16 (less optimal): 
`onednn_verbose,v1,primitive,exec,cpu,matmul,gemm:jit:bf16,undef,src:bf16::blocked:ab:768x1:f0 wei:bf16:a:blocked:ab::f0 dst:f32:a:blocked:ab::f0,,,14x64:64x142,0.118164`

```
#### **Without Strides**
```sh
ONEDNN_VERBOSE=all ./benchdnn --matmul --dt=bf16:bf16:f32  14x64:64x142

Maps to gemm:acl (optimal implementation): 
`onednn_verbose,v1,primitive,exec,cpu,matmul,gemm:acl,undef,src:bf16:a:blocked:ab::f0 wei:bf16:ap:blocked:BA8b4a::f0 dst:f32:a:blocked:ab::f0,,,14x64:64x142,0.150146`

```

I've done some relevant benchmarking for aarch64 for different models and shapes and concluded that not passing strides will improve the overall throughput.
Since it hasn’t been tested on all underlying architectures, I limited this to aarch64.

Comment (swolchok) [reply aten/src/ATen/native/mkldnn/Matmul.cpp:156]: well, I guess it must be OK as long as there is aarch64 test coverage for this. have you checked that it exists and we're running it?

**Review by malfet — COMMENTED** [review]
My two major concerns with this PR:
 - iDeep/MKLDNN suppose to provide an architecture agnostic primitives, this PR doesn't do a good job explaining why one codepath is needed for aarch64 and another for x86
 - not passing strides on aarch64 strongly hints at the correct issues, would love to see some explicit test in PR description that demonstrates strided tensor passed to this function and it works correctly (`test_matmul_lower_precision` is conditioned behind `is_supported`, which makes it hard to tell whether this is indeed tested or not)

**Comment by mingfeima** [comment]
@Valentine233 please help review this PR

**Comment by Valentine233** [comment]
The platform aarch64 can only work with MKLDNN, but not MKL. Without the MKLDNN path, aarch64 has to go into the fallback one. Although x86 can work with both MKL and MKLDNN, we still need to add a MKLDNN path for aarch64 here. Is it the case now? @aditew01

File: aten/src/ATen/native/CPUBlas.cpp [hunk 6]
```diff
@@ -417,13 +418,22 @@ void gemm(
       return;
    }
 #endif
+#if defined(__aarch64__) && AT_MKLDNN_ENABLED()
+// TODO: aditew01
```
[the remark this answers is not in the document]
Comment (aditew01) [reply aten/src/ATen/native/CPUBlas.cpp:421]: These are valid concerns. I've updated the description with the reasoning for why this explicit codepath change for aarch64 and x86. I hope this helps in making a case to why this specific case is divergent, making it not the cleanest solution but something which cannot be avoided immediately. 

TL;DR A thread which might help explain the different behaviour when strides are passed in oneDNN: https://github.com/pytorch/pytorch/pull/140159#discussion_r1983255391 

Also, I've attached a testcase explicitly which passes on NeoverseV1 instances. I hope this makes a case for correctness. I've benchmarked the accuracy for torchbench models with `precision=bf16`  and I don't see any accuracy failures with this change.

Updated the reasoning for TODO in the thread below.

**Review by nikhil-arm — APPROVED** [review]

File: aten/src/ATen/native/CPUBlas.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (aditew01) [reply aten/src/ATen/native/CPUBlas.cpp:426]: Bumping this up. Any thoughts / concerns @malfet ? Can we get this in Please.

File: aten/src/ATen/native/mkldnn/Matmul.cpp [hunk 7]
```diff
@@ -154,6 +153,98 @@ mkldnn_gemm(
...
```
[the remark this answers is not in the document]
Comment (aditew01) [reply aten/src/ATen/native/mkldnn/Matmul.cpp:156]: Sounds fair, ack. will push the req change.

File: aten/src/ATen/native/mkldnn/Matmul.cpp [hunk 8]
```diff
@@ -154,6 +153,98 @@ mkldnn_gemm(
...
+    ideep::tensor b({
+        /*sizes=*/{n, k},
+        idtype,
+        },
+      const_cast<scalar_t*>(b_data));
+    ideep::tensor c({
+        /*sizes=*/{n, m},
+        idtype,
+        },
+      c_data);
```
Comment (swolchok) [inline aten/src/ATen/native/mkldnn/Matmul.cpp:170]: this is new aarch64-specific no-strides behavior for f16x16->f16, f32xf32->f32, and bf16xbf16->bf16 GEMM (off-topic for this PR), and furthermore this behavior is *not* carried through to the almost-copy-paste of this code into the new bf16xbf16->f32 GEMM below.

after re-reading the PR description, I am confused; the description says this no-strides behavior is needed for exactly bf16xbf16->bf16 GEMM, but here you are applying it to the other two kernels I mentioned above as well. I would recommend the following as general improvements to the clarity of your PR:

1) move the explanation from the diff summary into the code as a comment so that readers of the code don't have to dig up this PR to understand what's going on.
2) either amend the explanation to include why we also need to cover f16xf16->f16 and f32xf32->f32, or fix the code if we don't actually need to do that. it might be easiest to fix the code if you created a pair of utility functions that created `ideep::tensor`s and fed them through `ideep::matmul_forward::compute`, one that used strides unconditionally (for the cases where that's correct) and one that omitted strides on aarch64 and otherwise called through to the first function (for use when necessary).
3) split the (apparently) unrelated enablement of bf16xbf16->f32 into a separate PR.

Comment (aditew01) [reply aten/src/ATen/native/mkldnn/Matmul.cpp:170]: Thanks for the review @swolchok.  I'll clear the clutter and separate this to 2 PR: 
1. enablement of bf16xbf16->f32. Certainly adding a function to create ideep::tensor will be a cleaner approach. I'll push that change with this.

2. fix strides passing when `idtype==bf16` for aarch64 when ACL is enabled. I'll make sure to add some explanation in the code comments to keep a track why the decision was made.

**Comment by aditew01** [comment]
@malfet @swolchok with this change, I am able to see a meaningful improvements in models using SDPA layers. 
Overall absolute execution time is improved (it effects the runtime for both Eager as well as Compiled mode, ie, Perf speedup is not the best measure to judge this.
 
 [PyTorch CI HUD - branch](https://hud.pytorch.org/benchmark/huggingface/inductor_no_cudagraphs?dashboard=torchinductor&startTime=Fri%2C%2028%20Mar%202025%2021%3A26%3A20%20GMT&stopTime=Fri%2C%2004%20Apr%202025%2020%3A26%3A20%20GMT&granularity=hour&mode=inference&dtype=bfloat16&deviceName=cpu%20(aarch64)&lBranch=adi/gemm_bf16f32&lCommit=d5aeab452e4b1f0580a4636b15a604c77a02c57b&rBranch=main&rCommit=bc72420bcb37390af3fced885e019903e6e425bd)
 
 Attaching a snippet for visibility : 

Screenshot 2025-04-04 at 21 33 51

 
Geometric mean speedup in HUD dashboard  : for Huggingface: `[0.48x → 0.58x]` and for Blueberries: `[0.88x → 1.13x]`

**Review by leslie-fang-intel — APPROVED** [review]
Overall LGTM. Some small comment and could you add the UT to cover the functionality of this new added function?

File: aten/src/ATen/native/mkldnn/Matmul.cpp [hunk 9]
```diff
@@ -155,35 +174,77 @@ mkldnn_gemm(
...
+    c.reorder_to(real_output);
+  }
+  return true;
+}
+
+template<typename scalar_t>
+inline typename std::enable_if_t<
+    std::is_same_v<scalar_t, c10::BFloat16>,
+    bool>
+mkldnn_gemm(
```
Comment (CaoE) [inline aten/src/ATen/native/mkldnn/Matmul.cpp:199]: Can this kernel be merged with the above kernel? It seems that most of the code is the same.

**Review by CaoE — APPROVED** [review]

File: aten/src/ATen/native/mkldnn/Matmul.cpp [hunk 10]
```diff
@@ -155,35 +174,77 @@ mkldnn_gemm(
...
+  if (transa != TransposeType::NoTranspose) {
+    std::swap(a_strides[0], a_strides[1]);
+  }
+  if (transb != TransposeType::NoTranspose) {
+    std::swap(b_strides[0], b_strides[1]);
+  }
+
+  auto idtype = ideep::tensor::data_type::bf16;
+  if constexpr (std::is_same_v<scalar_t, float>) {
+    idtype = ideep::tensor::data_type::f32;
```
[the remark this answers is not in the document]
Comment (aditew01) [reply aten/src/ATen/native/mkldnn/Matmul.cpp:232]: @leslie-fang-intel  ACK. This is redundant — I’ll remove it. Thanks!

Regarding the UT: the main goal is to support kernels used in scaled_dot_product_attention, such as Flash Attention and Efficient Attention. In these kernels, the operation q @ k.T = attn is performed where q and k are in bfloat16, and the output attn is in float32 to improve runtime performance.

I believe we already have both accuracy and operator tests that cover the bfloat16 path for SDPA. However, I don’t see a straightforward way to add a standalone unit test specifically for matmul with bfloat16 inputs and float32 accumulation. I'm open to suggestions on how can we add the test for this case.

File: aten/src/ATen/native/mkldnn/Matmul.cpp (hunk 9 above)
Comment (aditew01) [reply aten/src/ATen/native/mkldnn/Matmul.cpp:199]: Thanks for the review @CaoE . I believe It'll be a non-trivial change given the distinction is if we want the result to be in [float (usual accumulate dtype) ](https://github.com/pytorch/pytorch/blob/d5aeab452e4b1f0580a4636b15a604c77a02c57b/aten/src/ATen/native/mkldnn/Matmul.cpp#L206)/ scalar_t ( bfloat16 / float16) etc.   Please let me know your thoughts.

File: aten/src/ATen/native/mkldnn/Matmul.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (leslie-fang-intel) [reply aten/src/ATen/native/mkldnn/Matmul.cpp:232]: Thanks for the reply. Sound good to me if it has been covered by SDPA tests.

**Review by cfRod — APPROVED** [review]

File: aten/src/ATen/native/mkldnn/Matmul.cpp [hunk 11]
```diff
@@ -107,6 +107,25 @@ static bool use_mkldnn_bf32_matmul() {
...
 }
 
+// returns an ideep::tensor
+// - dims: shape e.g: {M,N}
+// - idtype: ideep data type e.g: (f32, bf16, f16)
+// - strides: Memory layout
+// - data: data pointer
+template <typename scalar_t>
+inline ideep::tensor make_ideep_tensor(
+    std::vector<int64_t> dims,
```
[the remark this answers is not in the document]
Comment (aditew01) [reply aten/src/ATen/native/mkldnn/Matmul.cpp:117]: Absolutely, it does create a temporary object. Typically, though, it’s called directly like this: make_ideep_tensor<scalar_t>({k, m}, ...).
An alternative could be explicitly assigning the dimensions to a named vector<int64_t> and passing it by const reference. Initially, I felt this might look less intuitive. Considering it’s just two integers, I assumed the overhead would be minimal.

I’d appreciate your suggestion on which approach you prefer.

**Review by fadara01 — APPROVED** [review]
LGTM!

**Review by malfet — APPROVED** [review]
